src/memray/_memray/macho_shenanigans.cpp

```diff
@@ -0,0 +1,145 @@
+#include <cstring>
+#include <dlfcn.h>
+
+#include "hooks.h"
+#include "linker_shenanigans.h"
+#include "logging.h"
+#include "macho_utils.h"
+
+#ifndef __APPLE__
+#    error "This file can only be compiled in MacOS systems"
+#endif
+
+namespace memray::linker {
+
+template<typename Hook>
+static void
+patch_symbol(
+        const Hook& hook,
+        typename Hook::signature_t intercept,
+        const char* symname,
+        void* addr,
+        vm_size_t section_size,
+        bool restore_original)
+{
+    vm_size_t trunc_address = trunc_page((vm_size_t)addr);
+    vm_size_t trunc_size = (vm_size_t)addr - trunc_address;
+
+    kern_return_t err = vm_protect(
+            mach_task_self(),
+            (uintptr_t)trunc_address,
+            section_size + trunc_size,
+            0,
+            VM_PROT_READ | VM_PROT_WRITE | VM_PROT_COPY);
+    if (err == KERN_SUCCESS) {
+        auto typedAddr = reinterpret_cast<typename Hook::signature_t*>(addr);
+        *typedAddr = restore_original ? hook.d_original : intercept;
+        LOG_IF_DEBUG(symname << " intercepted!");
+    } else {
+        LOG(ERROR) << "Failed to patch " << symname;
+    }
+}
+
+static void
+patch_symbols_in_section(
+        const section_t* section,
+        uintptr_t slide,
+        const DynamicInfoTable& dyninfo_table,
+        bool restore_original)
+{
+    auto symbol_addr_table = reinterpret_cast<void**>(slide + section->addr);
+    for (unsigned int i = 0; i < section->size / sizeof(void*); i++) {
+        const char* symbol_name = dyninfo_table.getSymbol(section->reserved1, i);
+        if (!symbol_name || !symbol_name[0] || !symbol_name[1]) {
```

Comment:
Why the `|| !symbol_name[1]` - seems to be ignoring symbols with single-character names, but why?

Oh wait, I think I figured it out - the first character is some dummy letter like `.` or `_` and that's why our strcmp below is ignoring it, right? If that's the case, should we be checking that it's exactly that character that we expect, instead of just anything but `'\0'`?